### PR TITLE
CorfuStore: Implementation of Streaming APIs

### PR DESCRIPTION
**Pull request by sneginhal**
## Overview

Description:

This is the first cut of the implementation of the streaming APIs of the CorfuStore.
This change has two commits, 
1. Sundar's API and stub change.
2. Change to implement the APIs.

I am still working on the unit tests.

Why should this be merged: 

Related issue(s) (if applicable): #<number>

## Checklist (Definition of Done):

- [ ] There are no TODOs left in the code
- [ ] [Coding conventions](https://github.com/CorfuDB/CorfuDB/wiki/Corfu-Style-Guidelines) (e.g. for logging, unit tests) have been followed
- [ ] Change is covered by automated tests
- [ ] Public API has Javadoc

### CODE REVIEW

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java [hunk 1]
```diff
@@ -57,4 +63,39 @@ public CorfuStreamEntry(K key, V payload, M metadata, long address, OperationTyp
         this.address = address;
         this.operation = operation;
     }
+
+    /**
+     * Convert a given SMREntry to CorfuStreamEntry.
+     */
+    public static <K extends Message, V extends Message, M extends Message>
+        CorfuStreamEntry<K, V, M> fromSMREntry(SMREntry entry, @Nonnull final Class<K> keyClass,
+            @Nonnull final Class<V> payloadClass, @Nonnull final Class<M> metadataClass) {
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: MetadataClass can be null.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java [hunk 2]
```diff
@@ -57,4 +63,39 @@ public CorfuStreamEntry(K key, V payload, M metadata, long address, OperationTyp
...
+        OperationType operationType = (entry.getSMRMethod().equals("put")) ? OperationType.UPDATE : OperationType.DELETE;
+        /*
+         * TODO[sneginhal]: Need a way to differentiate between update and create.
+         */
+        Object[] args = entry.getSMRArguments();
+
+        log.trace("Converting SMREntry at address {} to CorfuStreamEntry: Operation: {} Length of arguments: {}. " +
+                "({} -> {}, {})",
+                address, operationType, args.length, keyClass.getCanonicalName(), payloadClass.getCanonicalName(),
+                metadataClass.getCanonicalName());
```
Comment (zalokhan) [inline runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: Can throw NPE if metadataClass is null.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+                return false;
+            }
+
+            MultiObjectSMREntry multiObjectSMREntry = (MultiObjectSMREntry) logData.getPayload(runtime);
+            /*
+             * Build the CorfuStreamEntries from the MultiObjectSMREntry.
+             */
+            CorfuStreamEntries update = new CorfuStreamEntries(multiObjectSMREntry.getEntryMap().entrySet()
+                    .stream()
+                    .filter(e -> tablesOfInterest.keySet().contains(e.getKey()))
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: Done.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+
+    /**
+     * Acquire the StreamContext for polling.
+     */
+    public synchronized boolean acquire() {
+        /*
+         * If a TxPoller is already working on this StreamContext or an exception was encountered
+         * when polling return false.
+         */
+        if (txPollerInProgress || pollerException != null) {
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: Done

File: runtime/src/main/java/org/corfudb/runtime/collections/TxStreamer.java [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+
+    /**
+     * Executor service to run the txn stream poller.
+     */
+    private ExecutorService pollerExecutor = Executors.newWorkStealingPool(3);
+
+    /**
+     * Executor service to run the notifier.
+     */
+    private ExecutorService notifierExecutor = Executors.newWorkStealingPool(3);
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/TxStreamer.java]: Done

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java (hunk 1 above)
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: Done

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java (hunk 2 above)
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStreamEntry.java]: Done

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+            long startAddress) {
+        this.runtime = runtime;
+        this.listener = listener;
+        this.namespace = namespace;
+        this.tablesOfInterest = tableSchemas.stream().collect(Collectors.toMap(
+                        e -> runtime.getStreamID(runtime.getTableRegistry()
+                                .getFullyQualifiedTableName(namespace, e.getTableName())),
+                        e -> e));
+        this.startAddress = startAddress;
+        this.lastReadAddress = startAddress;
```
Comment (lixinchengdu) [inline runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: Is startAddress included in the stream?

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 7]
```diff
@@ -0,0 +1,268 @@
...
+        return false;
+    }
+
+    /**
+     * Deliver updates to the client.
+     *
+     * @param numUpdatesToDeliver Number of updates to deliver.
+     */
+    public void notifyClient(int numUpdatesToDeliver) {
+        if (notificationInProgress.compareAndSet(false, true)) {
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: If the compare fails, it means that another thread is already notifying this client, there is not much to be done.
This is where the client is being notified, the onNext() and onError() methods of the client StreamListener are called. We don't want the notifications to be delivered to the client in parallel, if another notification is in progress we don't want to block either, hence the guard.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    private final String namespace;
+
+    /**
+     * Tables of interest.
+     *
+     * Map of (Stream Id - TableSchema)
+     */
+    @Getter
+    private final Map<UUID,
+        TableSchema<? extends Message, ? extends Message, ? extends Message>> tablesOfInterest;
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: There will be a StreamContext per unique subscription via the subscribe() API.

[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: Maybe the name is confusing, does StreamSubscriptionContext seem better?

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 9]
```diff
@@ -0,0 +1,268 @@
...
+import org.corfudb.protocols.wireprotocol.ILogData;
+import org.corfudb.runtime.CorfuRuntime;
+
+/**
+ * This is the context of one client of the streaming API.
+ *
+ * Created by sneginhal on 10/22/2019.
+ */
+@Slf4j
+public class StreamContext {
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: Will do.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+                                .getFullyQualifiedTableName(namespace, e.getTableName())),
+                        e -> e));
+        this.startAddress = startAddress;
+        this.lastReadAddress = startAddress;
+    }
+
+    /**
+     * Acquire the StreamContext for polling.
+     */
+    public boolean acquire() {
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: ok

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+        synchronized (this) {
+            // If the lastAddress is not the same as lastReadAddress, something went wrong.
+            // Maybe multiple threads are trying to enqueue to the same StreamContext.
+            if (lastReadAddress != lastAddress) {
+                log.error("Incorrect address given to enqueueStreamEntry. Given {} Expected {}",
+                        lastAddress, lastReadAddress);
+                return false;
+            }
+
+            MultiObjectSMREntry multiObjSMREntry = (MultiObjectSMREntry) logData.getPayload(runtime);
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: ok.

File: runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java (hunk 6 above)
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/StreamContext.java]: No, start address is not included in the stream.

File: runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java [hunk 12]
```diff
@@ -151,4 +158,28 @@ public Query query(@Nonnull final String namespace) {
...
                 namespace);
     }
+
+    /**
+     * Subscribe to a specific a table in a namespace or the entire namespace.
+     * Objects returned will honor transactional boundaries
+     * @param streamListener - callback context.
+     * @param namespace - the CorfuStore namespace to subscribe to.
+     * @param tablesOfInterest - If specified, only updates from these tables will be returned.
+     * @param timestamp - If specified, all stream updates from this timestamp will be returned
```
[the remark this answers is not in the document]
Comment (sneginhal) [reply runtime/src/main/java/org/corfudb/runtime/collections/CorfuStore.java]: No, streaming tx after this timestamp till the unsubscribe is called.

File: runtime/src/main/java/org/corfudb/runtime/collections/TxnStreamingManager.java [hunk 13]
```diff
@@ -0,0 +1,210 @@
...
+
+import javax.annotation.Nonnull;
+
+import org.corfudb.runtime.CorfuRuntime;
+
+/**
+ * Created by sneginhal on 10/22/2019.
+ */
+@Slf4j
+public class TxnStreamingManager {
```
[the remark this answers is not in the document]
Comment (zalokhan) [reply runtime/src/main/java/org/corfudb/runtime/collections/TxnStreamingManager.java]: Changing it to TxnNotificationManager to keep it different from the underlying stream layer.

**Pull request merged**